알림 기능 구현

haha4825 opened the pull request:
## ✅ PR 체크리스트
- [x] Assignee를 지정했나요?
- [x] 병합 브랜치를 확인했나요?
- [x] 관련 이슈를 연결했나요?
- [x] 로컬 빌드에 성공했나요?

## 📝 변경 내용
- 댓글 알림
댓글 단 글에 댓글 달리면 알림

좋아요
본인이 쓴 글에 좋아요 달리면 알림

모임 참여
본인 주최 모임에 참여자 추가시 알림
모집 종료시 주최자 + 참여자에게 알림

## 📎 관련 이슈
- close #143

haha4825 committed:
[chore] gitignore firebase 파일 경로 추

haha4825 committed:
[chore] firebase 의존성 추가

haha4825 committed:
[feat] firebaseinitializer

haha4825 committed:
[feat] 알림 생성 메소드

haha4825 committed:
[feat] 댓글 레포지토리 코드 추가

haha4825 committed:
[chore] 설정 파일 firebase 경로 추가

haha4825 committed:
[feat] 도메인 별 알림 메소드 정의

haha4825 committed:
[feat] 도메인 별 알림 메소드 추

ruh0n commented on build.gradle:
```diff
@@ -36,15 +36,13 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-mail'
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
-    implementation 'com.amazonaws:aws-java-sdk-s3:1.12.791'
+    implementation 'com.google.firebase:firebase-admin:9.2.0'
```
최신버전은 [9.7.0](https://mvnrepository.com/artifact/com.google.firebase/firebase-admin)입니다.

구버전을 사용하시는 이유가 있나요?

haha4825 replied:
딱히 없어서 최신 버전으로 업데이트 했습니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/FCMService.java:
```diff
@@ -0,0 +1,56 @@
+                .putData("click_action", alarmDto.getUrl())
+                .build();
+            try {
+                FirebaseMessaging.getInstance().send(message);
```
sendAll함수가 있습니다. send는 동기 작업이라 후순위 유저는 알림이 매우 늦을 수 있을 것 같습니다.

haha4825 replied:
sendAll은 deprecate되어서 sendEachforMulticast를 사용하였습니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/FCMService.java:
```diff
@@ -0,0 +1,56 @@
+                FirebaseMessaging.getInstance().send(message);
+            } catch (FirebaseMessagingException e) {
+                failedTokens.add(token);
+                log.error("Failed to send notification to token: " + token.getTokenValue(), e);
```
https://firebase.google.com/docs/cloud-messaging/error-codes

모든 오류에서 토큰을 삭제하는 것은 잘못된 것 같습니다.

haha4825 replied:
토큰 값에 문제가 있을 때만 삭제하도록 변경하였습니다 감사합니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/dto/AlarmDto.java:
```diff
@@ -0,0 +1,21 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class AlarmDto {
```
알림 종류 정보도 있으면 좋을 것 같습니다

- POST_LIKED
- POST_COMMENTED
- ...

---

알림 고유 id도 있으면 좋을 것 같습니다.

알림이 중복으로 수신되는 등 여러 상황에서 유용하게 사용될 수 있습니다.

haha4825 replied:
흐음 알람 dto는 db에 저장되는 정보가 아니긴 한데 (다음과 같이 사용자에게 보낼 제목, 내용, 이미지, 클릭시 전환될 화면을 입력하는 용도로 저는 생각했습니다)
```java
public void sendParticipationAlarm(Participation participation) {

        LocalDate today = LocalDate.now();

        Member postOwner = participation.getMeetingPost().getMember();

        List<Token> tokens = tokenRepository.findByMemberInAndExpirationDateAfter(
            List.of(postOwner), today);
        if (tokens.isEmpty()) {
            return;
        }

        String postTitle = participation.getMeetingPost().getTitle();

        // 4) AlarmDto 생성 (타이틀/내용/이미지/URL은 서비스 정책에 맞게)
        AlarmDto alarmDto = AlarmDto.builder()
            .title("새로운 참여자가 생겼어요")
            .content("내가 작성한 " + postTitle + " 모임에 " + participation.getParticipant().getNickname()
                + "님이 참여합니다.")
            .url("/v1/boards/" + PostConstant.MEETING_POST_TYPE + "/posts/"
                + participation.getMeetingPost().getId()) // 프론트 라우팅에 맞게 수정
            .build();

        // 5) FCM 발사
        fcmService.sendByAlarm(alarmDto, tokens);
    }
    ```
그래도 필요할까요??

ruh0n replied:
@haha4825 

알림 정보는 알림 카테고리를 만들기 위해 사용할 수 있습니다.

os에서 기본 제공하는 사용자가 특정 카테고리의 알림만 차단할 수 있는 등의 기능입니다.

---

알림 id도 중복 알림을 피하기 위해서는 있는 것이 좋을 듯 합니다. 코드의 문제가 아니더라도 fcm에 의해 알림이 여러번 올 수 있는 상황도 있을 것 같습니다. 클라이언트에서는 id를 기준으로 필터링 하면 중복된 알림인지 알기가 쉽습니다. 물론이제 fcm Message에서 제공하는 androidconfig, apnsconfig까지 설정해서 보내주신다면 더 좋겠지만, 간단한 서비스에서는 과할 것 같아 id라도 있다면 좋을 것 같습니다.

앱의 관점에서 편의성을 위함이라 있으면 좋을 것 같다는 의견입니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/domain/Token.java:
```diff
@@ -0,0 +1,40 @@
+    private String tokenValue;
+
+    @Column(nullable = false)
+    private LocalDate expirationDate;
```
1. zone에 대한 정보가 없기때문에 LocalDate는 문제가 될 여지가 있어보입니다. 

2. 개인적으로는 만료일이 아니라 lastUsed와 같은 필드를 더 선호합니다.

haha4825 replied:
lastUsed로 변경하였습니다

haha4825 replied:
Instant로 변경했습니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/controller/FCMTokenController.java:
현재 구조는 토큰 해제 요청이 없어서 

사용자가 로그아웃 하더라도, 여전히 알림이 오는 문제가 발생할 수 있을 것 같습니다.

---

또한 새로운 token이 발급됐다 하더라도 기존 token이 바로 만료되는 것은 아닙니다.

즉 하나의 디바이스에 2개의 token이 동시에 사용될 수 있는 기간이 있을 수 있습니다. 그러면 알림이 중복으로 올 수 있습니다.

haha4825 replied:
1. 토큰 해제 요청 api를 만든다(로그아웃시 사용할 수 있다)
2. 토큰 등록시 기기이름을 같이 받도록 한다. (이미 사용자가 같은 기기명으로 토큰을 등록해두었다면 기존 토큰을 삭제하거나, 새로운 토큰 값으로 업데이트 한다)

=> 이렇게 기능을 업데이트 한다면 새로운 토큰 발급시 중복 문제와 로그아웃 모두 해결할 수 있을것 같습니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/domain/Token.java:
```diff
@@ -0,0 +1,40 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "member_id")
+    private Member member;
```
member를 이용한 조회가 많을 것 같은데 index 설정도 좋을 것 같습니다.

haha4825 replied:
적용했습니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/community/post/meetingpost/service/MeetingPostService.java:
```diff
@@ -202,15 +182,17 @@ public ParticipationResponse participate(Long postId, UsernamePasswordUserDetail
 
-        participationRepository.save(new Participation(participant, meetingPost));
-
+        alarmNotificationService.sendParticipationAlarm(participation);
```
알림이 lock이 적용된 transaction 내부에서 동기로 보내지는 것은  
동시에 요청한 다른 참여가 매우 늦어질 가능성이 커보입니다.

haha4825 replied:
트랜잭션 내부에서는 이벤트만 발행하고 후에 처리하는 방식으로 바꾸었습니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/community/like/service/DefaultLikeService.java:
```diff
@@ -21,27 +21,29 @@ public class DefaultLikeService implements LikeService {
+            .build();
 
         likeRepository.save(like);
+        alarmNotificationService.sendNewLikeAlarm(like);
```
like마다 알림이 보내진다면, 사용자가 장난삼아 like를 연타하면 어떻게 될까요? 😅

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/FCMTokenService.java:
```diff
@@ -0,0 +1,65 @@
+        log.info("오늘의 날짜 : " + now);
+
+        // 만료된 토큰을 가져옵니다.
+        List<Token> expiredTokens = tokenRepository.findByExpirationDate(now);
```
서버가 갑자기 죽어서 schedule이 실행되지 않는 날이 있다면 그 날의 token은 지워지지 않을 것 같습니다.

haha4825 replied:
두달 동안 사용하지 않은 토큰을 삭제하는 로직으로 바꾸어, 다시 서버가 띄워줬을 때 한번에 삭제되도록 변경하였습니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/FCMTokenService.java:
```diff
@@ -0,0 +1,65 @@
+    // 토픽 구독 - 토픽 하나씩
+
+    @Transactional
+    public void saveFCMToken(Member member, String fcmToken) {
```
tokenValue가 unique하기 때문에 사용자가 다른 계정으로 로그인 한다면 바로 exception이 발생할 것 같습니다.

haha4825 replied:
다른 계정으로 로그인한다면 멤버 객체가 달라 괜찮을거 같습니다.

ruh0n replied:
@haha4825 

https://github.com/park-su-park/Soomjae-back/pull/146/files#diff-032ea27a1d17d3f0661f44a29cf52aa4ca2d5c22611358b57d2f1f2b00eec2dbR31
다른 memberId에 같은 tokenvalue면 db scheme의 unique 위반인 것 같습니다.

haha4825 replied:
unique 조건을 없앴습니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/AlarmNotificationService.java:
```diff
@@ -0,0 +1,183 @@
+        receivers.addAll(commenters);
+
+        // 2) 중복 제거 + 이번에 댓글 쓴 사람 제외
+        receivers = receivers.stream()
+            .distinct()
+            .filter(m -> !m.getId().equals(writer.getId()))
+            .toList();
```
```suggestion
        receivers = commenters.stream()
            .distinct()
            .filter(m -> !m.getId().equals(writer.getId()))
            .toList();
```

commenters로 바로 filter하면 안되나요? 🤔

haha4825 replied:
본인이 작성한 댓글은 알림을 받지 않도록 하기 위해서입니다!

ruh0n replied:
@haha4825 receivers.addAll한 이후 다시 필터링 하는 것이 문제라고 생각되는 점입니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/AlarmNotificationService.java:
```diff
@@ -0,0 +1,183 @@
+
+        // 2) 중복 제거 + 이번에 댓글 쓴 사람 제외
+        receivers = receivers.stream()
+            .distinct()
```
unique는 db에서 먼저 처리하고 오는 것이 좋을 듯 합니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/AlarmNotificationService.java:
```diff
@@ -0,0 +1,183 @@
+
+        List<Member> receivers = new ArrayList<>();
+
+        // 댓글 단 사람들 추가
+        List<Member> commenters = commentRepository.findMembersByPost(comment.getPostType(),
+            comment.getPostId());
```
게시글 작성자에게는 안 보내주나요?

haha4825 replied:
추가했습니다

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/service/AlarmNotificationService.java:
```diff
@@ -0,0 +1,183 @@
+        // 3) 멤버들의 유효 토큰 조회
+        LocalDate today = LocalDate.now();
+
+        List<Token> tokens = tokenRepository.findByMemberInAndExpirationDateAfter(receivers, today);
```
멤버를 불러온 후 토큰을 가져오면 2번의 query 요청이 발생하지만,

join으로 한번에 가져온다면 1번으로도 충분할 것 같습니다.

ruh0n commented on src/main/java/com/parksupark/soomjae/server/fcm/FCMInitializer.java:
```diff
@@ -0,0 +1,37 @@
+    @Value("${fcm.firebase_config_path}")
+    private String firebaseConfigPath;
+
+    @PostConstruct // 빈 객체가 생성되고 의존성 주입이 완료된 후에 초기화가 실행될 수 있도록 @PostConstruct 설정
```
잘 몰라서 그러는데 PostConstruct를 사용해야 할 필요가 있나요?

ruh0n reviewed (commented):
고생하셨습니다 👍

haha4825 committed:
[chore] firebase 의존성 9.2.0 -> 9.7.0

haha4825 committed:
[fix] 알림을 보낼 때 모든 예외 코드에서 토큰을 삭제하는 것이 아닌 토큰 값에 문제가 있을 때만 db에서 토큰 삭제

haha4825 commented:
> 고생하셨습니다 👍

폭풍 리뷰 감사합니다. 모두 반영하겠습니다!

rlatnfla commented on src/main/java/com/parksupark/soomjae/server/community/comment/service/CommentService.java:
```diff
@@ -24,52 +25,55 @@ public class CommentService {
         validatePost(postType, postId);
         Comment comment = commentRepository.findByIdAndDeletedTimeIsNull(commentId)
-                .orElseThrow(() -> new IllegalStateException("해당 Id를 가진 comment가 존재하지 않습니다."));
+            .orElseThrow(() -> new IllegalStateException("해당 Id를 가진 comment가 존재하지 않습니다."));
```
`ResourceNotFoundException` 클래스를 활용하시면 좋을 것 같습니다!

haha4825 committed:
[fix] 토큰 삭제 로직을 마지막 사용시점으로 변경(두달 간 사용하지 않으면 삭제)

haha4825 committed:
[fix] PR리뷰 반영

최근 사용일을 기준으로 토큰 생명관리
fcm token 삭제 api추가
토큰 등록시 기기 명을 받도록 변 (같은 기기, 사용자로 토큰이 등록되어있다면 기존 새로운 토큰을 등록하는 것이 아닌 기존의 토큰값을 업데이트한다)
토큰에 멤버로 인덱스 추가

ruh0n commented:
> > 고생하셨습니다 👍
> 
> 폭풍 리뷰 감사합니다. 모두 반영하겠습니다!

@haha4825 모두 반영하실 필요는 없습니다! 

명확한 이유가 있으시다면 오히려 반영하지 않으시는 것이 더 좋습니다!

haha4825 committed:
[fix] 비동기 방식으로 참여 알림 전송

haha4825 committed:
[fix] tokenValue 유니크 제약 조건 없앰

haha4825 committed:
Merge branch 'dev' into taewon/feat/#143/alarm

haha4825 committed:
[fix] 충돌해결하며 생긴 코드 원복

haha4825 committed:
[style] checkstyle

haha4825 committed:
[test] AlarmNotificaiton 클래스 Mock 주입

haha4825 committed:
[test] AlarmNotificaiton 클래스 Mock 주입
